### PR TITLE
[Core] Simplify the Dp padding/should ubatch coordination logic

### PR DESCRIPTION
**Pull request by SageMoore**
## Purpose
This PR attempts to improve the robustness and simplify the cross-DP coordination logic. Especially after adding DBO, this logic has become incredibly complicated and brittle.

There are several pieces of data that need to be synchronized by DP Ranks.
1. The number of tokens after padding.
2. Is this rank attempting to run with DBO

This PR attempts to consolidate all of this DP coordination logic into a single function `coordinate_batch_across_dp`. This function will be called once per-rank per-iteration of the model runner.

## Test Plan

lm_eval

## Test Result
`VLLM_ALL2ALL_BACKEND=deepep_low_latency vllm serve --model="deepseek-ai/DeepSeek-V2-Lite" --trust-remote-code --data-parallel-size 2 --enable-expert-parallel `

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.3567|±  |0.0277|
|     |       |strict-match    |     5|exact_match|↑  |0.3533|±  |0.0276|
```

`VLLM_ALL2ALL_BACKEND=deepep_low_latency vllm serve --model="deepseek-ai/DeepSeek-V2-Lite" --trust-remote-code --data-parallel-size 2 --enable-expert-parallel --enable-dbo --dbo-decode-token-threshold 4 `

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.3633|±  |0.0278|
|     |       |strict-match    |     5|exact_match|↑  |0.3600|±  |0.0278|
```

### CODE REVIEW

File: vllm/v1/worker/dp_utils.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+def _get_device_and_group():
+    from vllm.distributed.parallel_state import get_dp_group
+    device = current_platform.device_type
+    group = get_dp_group().device_group
+
+    # Transfering this tensor from GPU to CPU will introduce a GPU sync
+    # point that could adversely affect performance of vllm with asynch
+    # scheduling. This environment variable exists to quickly disable
+    # this optimization if we run into this case.
+    if envs.VLLM_DISABLE_NCCL_FOR_DP_SYNCHRONIZATION:
```
Comment (LucasWilkinson) [inline vllm/v1/worker/dp_utils.py]: can we make this config option and add it to the CLI? I think it will make it easier to override if we want to turn it on in async scheduling case

File: vllm/v1/worker/dp_utils.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        logger.info_once(
+            "Using CPU all reduce to syncronize DP padding between ranks.")
+        device = "cpu"
+        group = get_dp_group().cpu_group
+    return device, group
+
+
+def _run_ar(should_ubatch: bool, orig_num_tokens_per_ubatch: int,
+            padded_num_tokens_per_ubatch: int, dp_size: int,
```
Comment (LucasWilkinson) [inline vllm/v1/worker/dp_utils.py]: can we get `dp_size` directly from `group`?

File: vllm/v1/worker/dp_utils.py (hunk 1 above)
Comment (SageMoore) [reply vllm/v1/worker/dp_utils.py]: Now that we have a clearer picture of how poorly this impacts async scheduling, I think a CLI option makes a lot of sense.

File: vllm/v1/worker/dp_utils.py (hunk 2 above)
Comment (SageMoore) [reply vllm/v1/worker/dp_utils.py]: Looks like we can get size and rank from the `GroupCoordinator`

**Review by LucasWilkinson — APPROVED** [review]
LGTM! Thanks of the clean-up!

File: tests/quantization/test_blackwell_moe.py [hunk 3]
```diff
@@ -15,6 +15,9 @@
         "This test only runs on Blackwell GPUs (SM100).", allow_module_level=True
     )
 
+# Make sure TRTLLM attention is available
+os.environ["VLLM_HAS_FLASHINFER_CUBIN"] = "1"
+# Set compilation threads to 16 to speed up startup
```
Comment (LucasWilkinson) [inline tests/quantization/test_blackwell_moe.py]: remove unrelated change

File: vllm/v1/worker/gpu_ubatch_wrapper.py [hunk 4]
```diff
@@ -409,6 +410,18 @@ def __call__(self, *args, **kwargs):
...
+            ubatch_slices[0].token_slice.stop - ubatch_slices[0].token_slice.start
+        )
+        dp_size = self.vllm_config.parallel_config.data_parallel_size
+        sliced_num_tokens_across_dp = torch.tensor(
+            [num_tokens_per_ubatch] * dp_size, device="cpu", dtype=torch.int32
+        )
+        sliced_dp_metadata = DPMetadata.make(
+            self.vllm_config.parallel_config,
+            num_tokens_per_ubatch,
+            sliced_num_tokens_across_dp,
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu_ubatch_wrapper.py]: this naming confuses me a bit; what makes it "sliced"

Comment (SageMoore) [reply vllm/v1/worker/gpu_ubatch_wrapper.py]: "ubatched" is probably a better word.

Comment (LucasWilkinson) [reply vllm/v1/worker/gpu_ubatch_wrapper.py]: or `ubatch_dp_metadata`? since it specific to a single ubatch if im understanding this correctly?

**Review by tlrmchlsmth — COMMENTED** [review]
FYI: this PR broke TP+DP
